File: functions/_lib/adapters/kaobaAdapter.js
/**
 * KaobaReplyAdapter — 応答生成のみ差し替え
 *
 * 契約（expect-kaoba/1.0）・POST /api/kaoba/chat は変更しない。
 *
 * プロバイダ優先順位:
 *   1. python … AI_BASE_URL → POST /v1/kaoba/chat（将来 LLM 実装）
 *   2. rule  … generateKaobaReply（現行デフォルト）
 *
 * env.KAOBA_PROVIDER = "python" | "rule" | "auto"（default: auto）
 *   auto = python 成功時は python、失敗時 rule
 */
import { aiFetch } from "../aiProxy.js";
import { getEnv } from "../env.js";
import {
  generateKaobaReply as ruleGenerateKaobaReply,
  loadKaobaRefs,
  normalizeKaobaResponse,
} from "../kaobaDomain.js";
import { adaptAnalysisGet } from "./analysisAdapter.js";
import { adaptPredictionGet } from "./predictionAdapter.js";

function providerMode(context) {
  const raw = String((getEnv(context).KAOBA_PROVIDER || "auto")).toLowerCase();
  if (raw === "python" || raw === "llm" || raw === "rule") return raw === "llm" ? "python" : raw;
  return "auto";
}

/**
 * race_id 参照を Adapter 経由で解決（Mock 直読みをやめる）
 * Prediction / Analysis が実AIになっても Kaoba は同じ経路で追従する。
 */
export async function loadKaobaRefsViaAdapters(context, raceId) {
  if (!raceId) {
    return { race_id: null, bundle: null, analysis: null };
  }

  const [pred, anal] = await Promise.all([
    adaptPredictionGet(context, raceId),
    adaptAnalysisGet(context, raceId),
  ]);

  const bundle = pred && pred.ok ? pred.bundle : null;
  const analysis = anal && anal.ok ? anal.analysis : null;

  return {
    race_id: raceId,
    bundle,
    analysis,
    ref_sources: {
      prediction: pred && pred.ok ? pred.provider : "none",
      analysis: anal && anal.ok ? anal.provider : "none",
    },
  };
}

async function generateViaPython(context, input) {
  const proxied = await aiFetch(context, "/v1/kaoba/chat", {
    method: "POST",
    body: JSON.stringify({
      message: input.message,
      history: input.history || [],
      race_id: input.race_id || null,
      context: input.context || {},
    }),
  });
  if (proxied && proxied instanceof Response) {
    return { ok: false, errorResponse: proxied };
  }
  if (!proxied || !proxied.ok) return null;
  const data = proxied.payload.data != null ? proxied.payload.data : proxied.payload;
  return {
    ok: true,
    response: normalizeKaobaResponse(data, input.race_id || null),
    source: "single-ai",
    provider: "python",
  };
}

/** ルールエンジン（現行）。差し替え時はこの関数参照だけを変える。 */
export function generateViaRule(input) {
  const response = ruleGenerateKaobaReply(input);
  return {
    ok: true,
    response,
    source: "mock",
    provider: "rule",
  };
}

/**
 * 唯一の公開生成入口 — ルートはこれを呼ぶ
 * @returns {Promise<{ ok: true, response, source, provider }>}
 *
 * auto / python: Python 成功時のみ python。失敗（不通・502 等）は rule へフォールバック。
 * （errorResponse をそのまま返さない — 契約レスポンスを維持）
 */
export async function adaptKaobaChat(context, body) {
  const message = String((body && body.message) || "").trim();
  const raceId = String(
    (body && (body.race_id || (body.context && body.context.bundle_ref))) || ""
  ).trim() || null;
  const history = Array.isArray(body.history) ? body.history : [];
  const contextUi = (body && body.context) || {};

  const refs = await loadKaobaRefsViaAdapters(context, raceId);
  const input = {
    message,
    history,
    race_id: raceId,
    context: contextUi,
    refs,
  };

  const mode = providerMode(context);

  if (mode === "rule") {
    return generateViaRule(input);
  }

  if (mode === "python" || mode === "auto") {
    const fromPy = await generateViaPython(context, input);
    if (fromPy && fromPy.ok) return fromPy;
    // python 不通 / Network error / 非OK → rule（auto・python 強制とも契約応答を維持）
    const fallback = generateViaRule(input);
    fallback.provider = mode === "python" ? "rule-fallback" : "rule";
    return fallback;
  }

  return generateViaRule(input);
}

export const KaobaAdapter = {
  chat: adaptKaobaChat,
  generateViaRule,
  loadRefs: loadKaobaRefsViaAdapters,
  /** 互換: 旧 generateKaobaReply 名 */
  generateKaobaReply: ruleGenerateKaobaReply,
  /** 旧 loadKaobaRefs は adapters 経由へ誘導 */
  loadKaobaRefs,
};

File: services/win5-ai/app/engine/adapters/kaoba_adapter.py
"""
KaobaAdapter (Python)

契約: expect-kaoba/1.0（変更禁止）
現行: domains.kaoba_reply（rule）
移行: LlmKaobaSource.generate を実装し KAOBA_PROVIDER=llm で切替
"""
from __future__ import annotations

import os
from typing import Any

from .. import domains
from .analysis_adapter import get_analysis
from .prediction_adapter import get_bundle


class RuleKaobaSource:
    """現行ルールエンジン。"""

    def generate(self, body: dict[str, Any]) -> dict[str, Any]:
        payload = dict(body if isinstance(body, dict) else {})
        race_id = str(payload.get("race_id") or "")
        if not race_id:
            ctx = payload.get("context") if isinstance(payload.get("context"), dict) else {}
            race_id = str(ctx.get("bundle_ref") or "")
            if race_id:
                payload["race_id"] = race_id
        if race_id and "_bundle" not in payload:
            try:
                bundle = get_bundle(race_id)
                if isinstance(bundle, dict) and bundle:
                    payload["_bundle"] = bundle
            except Exception:
                pass
        return domains.kaoba_reply(payload)


class LlmKaobaSource:
    """
    LLM 用プレースホルダ。
    実装時: body + optional bundle/analysis context → KaobaChatResponse。
    未実装時は rule へフォールバック。
    """

    def __init__(self, fallback: RuleKaobaSource | None = None) -> None:
        self._fallback = fallback or RuleKaobaSource()

    def generate(self, body: dict[str, Any]) -> dict[str, Any]:
        # TODO: call LLM with refs from get_bundle / get_analysis
        race_id = str((body or {}).get("race_id") or "")
        if race_id:
            _ = get_bundle(race_id)
            _ = get_analysis(race_id)
        out = self._fallback.generate(body)
        out = dict(out)
        out["provider"] = "rule-fallback"
        return out


def _provider() -> str:
    return (os.environ.get("KAOBA_PROVIDER") or "rule").lower()


class KaobaAdapter:
    def __init__(self) -> None:
        rule = RuleKaobaSource()
        mode = _provider()
        self._source = LlmKaobaSource(rule) if mode in ("llm", "python", "auto") else rule

    def generate(self, body: dict[str, Any]) -> dict[str, Any]:
        return self._source.generate(body)


_adapter = KaobaAdapter()


def generate_reply(body: dict[str, Any]) -> dict[str, Any]:
    """公開入口 — main.py はこれを呼ぶ（domains.kaoba_reply 直呼びをやめる）。"""
    return _adapter.generate(body)
